Improve performance of string search.

Rot127 opened the pull request:
# WIP

<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Depends on:
- https://github.com/rizinorg/rizin-testbins/pull/196

Measures taken:

- Check decoded strings partially gainst the patterns. This means we can returns early without decoding many invalid strings.
- Use a O(1) is_valid_unicode_point function, replacing the O(log n) check.

### TODO

- [x] Native UTF-16/32 regex search.
- [x] Parial matching while decoding.
	- [x] JIT compiler must be cloned if used by multiple threads.
	- [x] Cloning and recompiling JIT might be overkill for literal regex patterns.
	      Make an exception for them?
- [x] ! utf8_to_mem_offset_map is way way too expensive and a stupid idea. Do it at the end somehow.
- [x] Alignment defined by user is ignored in scan_strings
- [x] Make fast code point decode optional
- [ ] Rename some variables for clarity
- [x] Move `hit_type` string description onto stack.
- [ ] Add note about foreign endian search in command.
- [ ] Remove `debaseb64` option (in extra PR due to project version increment)
- [x] Don't alloc matches in `match_first` before needed.
- [ ] Open issue about more performance improvements (see: https://pcre2project.github.io/pcre2/doc/pcre2jit/).
  - Allocate JIT stacks for each thread before search begins, pass the matching context in each worker. This would remove the current cloning of the regex objects for each worker.
  - Always use `pcre2_jit_match()` to ignore sanity checks of `pcre2_match()`.
  - [Medium fast cross endian regex matching of UTF16/32](https://github.com/PCRE2Project/pcre2/issues/763#issuecomment-3092352856).
- [ ] Document the internal string search options.
- [ ] NULL checks to multi API functions
- [ ] Remove `RZ_STRING_ENC_BASE64`

**MUST CHECK**

- [x]  ! Byte search UTF_NOCHECK can lead to UB. For utf8 strings. But byte search explicitly passes those. It needs to match extended unicode.
- [ ] Remove `match_inverse` because it is unused and not well enough defined for files with  binary data in it.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

Rot127 commented:
Preliminary performance (files are strings mixed with data):

```
ll test_strings_ja.utf-*
-rw-r--r--. 1 user user 543M Jul 12 07:53 test_strings_ja.utf-16-le
-rw-r--r--. 1 user user 1.1G Jul 12 07:53 test_strings_ja.utf-32-be
-rw-r--r--. 1 user user 663M Jul 12 07:53 test_strings_ja.utf-8
```

```
> time rizin -n -qc "e str.encoding=utf16le ; /z '明らかにする リハビリ 探査 メニュー ホイール '" test_strings_ja.utf-16-le 
0x000343af 50 hit.string.utf16le.0

real	1m33.091s
user	5m17.410s
sys	0m8.139s

> time rizin -n -qc "e str.encoding=utf8 ; /z '明らかにする リハビリ 探査 メニュー ホイール '" test_strings_ja.utf-8 
0x0003fde4 65 hit.string.utf8.0

real	0m5.471s
user	0m19.625s
sys	0m0.175s

time rizin -n -qc "e str.encoding=utf32be ; /z '明らかにする リハビリ 探査 メニュー ホイール '" test_strings_ja.utf-32-be 
0x00062d99 100 hit.string.utf32be.0

real	1m6.058s
user	3m50.468s
sys	0m10.794s
```

This is still way too slow for large scale automation. But ok for "one user works on a binary" tasks.

wargio commented on librz/util/unicode.c:
```diff
@@ -308,6 +308,35 @@ RZ_API bool rz_unicode_code_point_is_legal_decode(const RzCodePoint c) {
+	       || RZ_BETWEEN(c, 0x2fa1e, 0x2ffff)
+	       || RZ_BETWEEN(c, 0x1dab0, 0x1deff)
+	       || RZ_BETWEEN(c, 0x1fbfa, 0x1ffff)
+				 || rz_unicode_code_point_is_surrogate(c);
```
format the spaces.

Rot127 committed:
Make checks against defined Unicode points optional

Rot127 committed:
Allow to decode UTF-16 without writing the result.

Rot127 committed:
Remove PCRE2_NO_UTF_CHECK as default, since it can lead to undefined behavior.

Rot127 committed:
First refactor regex to support utf16 and utf32

Rot127 committed:
Add UTF-16-BE encoding function

Rot127 committed:
Add UTF-8 counting helper functions.

- One for counting the number of Unicode code points.
- The other to get the number of bytes required to represent the given UTF-8 string in UTF-16.

Rot127 committed:
Remove unused code

Rot127 committed:
Add UTF-8 to UTF-16 conversion function.

Rot127 committed:
Add type annotations

Rot127 committed:
Implement utf8 to utf32 string conversion

Rot127 committed:
Add UTF-16/32 versions of all other necessary regex functions for str search.

Rot127 committed:
Another regex refactor for utf16/32

Rot127 committed:
Add UTF16/32 regex matching tests.

Rot127 committed:
Implement still segfaulting (possibly JIT double usage) regex search.

notxvilka commented on librz/util/utf16.c:
```diff
@@ -81,27 +94,34 @@ RZ_API size_t rz_utf16le_decode(RZ_NONNULL const ut8 *buf, size_t buf_len, RZ_NO
 	if (codepoint < 0x10000) {
-		buf[0] = codepoint & 0xff;
-		buf[1] = codepoint >> 8 & 0xff;
+		buf[big_endian ? 1 : 0] = codepoint & 0xff;
```
In situations like this (here and below) to avoid inlining conditions in each "move"/"store" it's better to pre-calculate indices once. Especially for hot paths.

Rot127 replied:
Yes, I thought the compiler will optimize it this way.
But better not trust in this.

notxvilka commented on librz/util/utf16.c:
```diff
@@ -110,10 +130,10 @@ RZ_API size_t rz_utf16le_encode(RZ_NONNULL RZ_OUT ut8 *buf, RzCodePoint codepoin
-	buf[1] = high >> 8 & 0xff;
-	buf[2] = low & 0xff;
-	buf[3] = low >> 8 & 0xff;
+	buf[big_endian ? 1 : 0] = high & 0xff;
```
Same here

notxvilka commented on librz/util/utf32.c:
```diff
@@ -77,3 +82,24 @@ RZ_API bool rz_utf32_valid_code_point(RZ_NONNULL const ut8 *buf, size_t buf_len,
+	if (ucp > RZ_UNICODE_LAST_CODE_POINT || rz_unicode_code_point_is_surrogate(ucp)) {
+		return 0;
+	}
+	buf[big_endian ? 3 : 0] = ucp & 0xff;
```
Same here

notxvilka reviewed (changes requested):
Just a minor performance suggestion

Rot127 committed:
Duplicate match_first functions to reduce necessary branch predictions.

Rot127 committed:
Reduce number of required branches for encoding UTF16/32 to one.

Rot127 committed:
Duplicate match_all_internal functions to reduce necessary branch predictions.

Rot127 committed:
Fix too early free

Rot127 committed:
Only allocate match vector when needed.

Rot127 committed:
Fix: use code point size of buffer.

Rot127 committed:
Add missing return

Rot127 committed:
Normalize pointers to UTF16/32 strings to use proper code point with

Rot127 committed:
Also replace spaces with in utf16/32

Rot127 committed:
Add an additional host endian tests for UTF16/32 string encodings.

Rot127 committed:
Ensure thread savety.

JIT compiled patterns need to be owned by a single thread.
For the search we need to clone it.

JIT matching structures are optionally cloned as well.

Rot127 committed:
Enforce NO_UTF_CHECK in regex search.

This improves performance and currently is
default because we always match on binary data.

Rot127 committed:
Fix matching of UTF strings which are not suported by direct buffer matching.

PCRE2 only supports matching against memory which is aligned
to a code point width of the encoding.

These changes prevent taking the fast (direct matching with PCRE2) path
and use the slow string search path if the alignment doesn't match
the UTF string encoding.

To not complicate the change and additional alignment member
is added to each searched string in the search collection.

Rot127 committed:
Create search hit description on the stack

Rot127 committed:
Unset complete JIT matching if user provided custom jflags.

Rot127 committed:
Document what passing NULL to copy function pointers does.

Rot127 committed:
Replace the retarded idea of tracking offsets with a hashmap with a linear buffer.

This improves performance something like 10x.

Rot127 committed:
Remove const for the non-JIT builds.

Rot127 committed:
Remove additional flags for skip checking.

It is not needed because each decoded character is checked for printablity below anyways.

Rot127 committed:
Enfore no setup of IO mem with 0xff

Rot127 committed:
Fix: Set JIT complete flag for multi regex patterns

Rot127 commented:
**Methods**

| Speed | configuration | String encoding guessed? | Search strategy |
|-------|---------------|--------------------------|--|
| Fast | Encoding is specified via `e str.encoding` --- Encoding is in host's endianness --- Address alignment is the same as code point width (1/2/4 bytes) | No | Pass buffer directly to regex matching |
| Slow | Encoding is specified via `e str.encoding --- Encoding is **not** in host's endianness --- **and/or** --- Address alignment is **not** the same as code point width (1/2/4 bytes)  | No | 1. Scans the buffer for all strings of the specified encoding --- 2. Matches detected strings. |
| Very Slow | Doesn't matter | Yes | Does a lot before doing the `Slow` method. |
| Normal (`grep -acE "pattern" utf8-file`) | - | No | Linear matching of buffer I guess. |

**Stats**

Example command:
```
time rizin -n -qc "e search.align=2 ; e str.encoding=utf16le ; /z '些 公司 任何....可' r" test/bins/test_strings_zh.utf-16-le
```

**Files**
```
Files contain strings of given encoding, mixed with random data.
Strings are always aligned to code point widths.
```

**Runtimes**

| File / method                    | Normal     | Fast (Fast without JIT)    | Slow                    | Very Slow |
|--|--|--|--|--|
| test_strings_zh.utf-32-le (1.1G) | -          | 0m0.297s (0m2.024s) | 0m17.894s          | 2m18.581s (fails to find string) |
| test_strings_zh.utf-32-be (1.1G) | - | Equivalent to utf-32-le on big endian machines     | 0m17.768s   | 2m34.502s (Fails to detect string |
| test_strings_zh.utf-16-le (552M) | -          | 0m0.417s (0m1.241s) | 0m45.271s  | 2m17.124s |
| test_strings_zh.utf-16-be (552M) | -          | Equivalent to utf-16-le on big endian machines | 0m51.738s | 4m10.638s |
| test_strings_zh.utf-8 (630M)     | 0m0.525s   | 0m0.421s (0m2.237s) | Always `Fast` `str.encoding=utf8`, otherwise always `Very Slow`           | 2m44.419s |

It goes without saying that these numbers are not averaged or anything.
Especially with the fast methods I sometimes get +0.2s.
This runs on a VM with 13GB RAM, 4 VCPUs (1.1-4.3 GHz).
But for the slow methods less than 4GB of RAM is used.

~~Still need to check what is going on with the OOM kills.~~ Was a leak.

Rot127 committed:
Fix heap.

Rot127 committed:
Add note about worsed performance path.

notxvilka reviewed (changes requested):
Could you please also add `librz/search/README.md`? To describe the "module" and the code structure (see `librz/README.md`). You could use `librz/util/README.md` or `librz/io/README.md` as an example, though something slightly more verbose would be better. For both people and AI assistants.

Rot127 committed:
Fix unit test with string terminated by undefined code point.

Rot127 committed:
Run clang-format

Rot127 committed:
Fix order of arguments

Rot127 committed:
Add warning about string search with encoding=guess to tests.

Rot127 committed:
Fix string lengths, they no longer count the final invalid code point.

Rot127 committed:
Fix endian macro on Windows

Rot127 committed:
Fix NULL dereference

Rot127 committed:
Fix number tests

Rot127 committed:
Use endianness check not dependent on stdbit

Rot127 committed:
Fix type annotations.

Rot127 committed:
Reintroduce rz_str_len_utf8char

wargio commented on librz/include/rz_platform.h:
```diff
@@ -13,6 +13,20 @@
 extern "C" {
 #endif
 
+#if defined(__BYTE_ORDER__)
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define RZ_HOST_BIG_ENDIAN false
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#define RZ_HOST_BIG_ENDIAN true
+#else
+#error "Unsupported endianness"
+#endif
+#else
+// Hacky version to determine endianness
+static const ut32 rz_endianness_one = 1;
+#define RZ_HOST_BIG_ENDIAN !(*((char *)&(rz_endianness_one)) == 1)
+#endif
```
i'm quite sure this is available in `rz_types.h` as `RZ_SYS_ENDIAN` when `0` is LE, when `1` is BE

Rot127 replied:
Oh thanks! Overlooked this.
But most of them are set to `RZ_SYS_ENDIAN_NONE (0)`:
```
#define RZ_SYS_ENDIAN_NONE   0
#define RZ_SYS_ENDIAN_LITTLE 1
#define RZ_SYS_ENDIAN_BIG    2
#define RZ_SYS_ENDIAN_BI     3
```

Should I replace the macro then?

Rot127 replied:
Or we switch the macro name to `RZ_BUILD_TIME_ENDIANNESS`

Rot127 replied:
Never mind, I'll update it.

wargio replied:
yeah, that is indeed wrong. now i understand why we had issues with @ahmed-kamal2004 PR on libmagic

wargio commented on librz/search/string_search.c:
```diff
@@ -215,11 +297,56 @@ RZ_API bool rz_search_collection_string_add(RZ_NONNULL RzSearchCollection *col,
+		if (log_suggestions) {
+			RZ_LOG_WARN("The string encoding for the search is set to \"guess\".\n"
+				    "The search will consume vastly more resources and the guessing is unreliable.\n"
+				    "You can set a specific encoding with 'e str.encoding'.\n");
```
```suggestion
				    "You can set a specific encoding with 'e str.encoding=<format>'.\n");
```

wargio commented on librz/util/str.c:
```diff
@@ -4331,3 +4435,59 @@ RZ_API const char *rz_str_indent(int indent) {
+	switch (enc) {
+	default:
+	case RZ_STRING_ENC_GUESS:
+	case RZ_STRING_ENC_BASE64:
```
shouldn't `RZ_STRING_ENC_BASE64` be `1` ?

Rot127 replied:
`RZ_STRING_ENC_BASE64` shouldn't be part of the enum actually. Base64 is not a string encoding like Unicode or the IBM encodings.
Added the remove it steps to the todo above.

wargio reviewed (approved):
overall looks good. just some nitpicks here and there.

Rot127 committed:
Fix command description.

Rot127 committed:
Fix and unify RZ_SYS_ENDIAN macros.

- Don't allow unhandled architectures anymore.
- Check endianness for Sparc and PPC using non GCC/Clang compilers.
- Fix several endianness checks using the value instead of the macros.

Rot127 commented:
Pushed to `dist-str-search-perf` to check the endianness stuff

Rot127 committed:
Fix tests

notxvilka commented on librz/include/rz_types.h:
```diff
@@ -400,119 +400,150 @@ static inline void *rz_new_copy(int size, const void *data) {
+			#error "Unsupported endianness"
+		#endif
+	#else
+		#define RZ_SYS_ENDIAN ((*((char *)&(rz_endianness_one)) == 1) ? RZ_SYS_ENDIAN_LITTLE : RZ_SYS_ENDIAN_BIG)
```
In the future we should move this check to the Meson (Muon) level.

notxvilka commented on librz/include/rz_util/rz_str.h:
````diff
@@ -281,10 +290,44 @@ RZ_API RZ_OWN char *rz_str_stringify_raw_buffer(RzStrStringifyOpt *option, RZ_NU
+ * Examples:
+ *
+ * ```c
+ * // IBM290 character width is always one byte, but the equivalent Japanes
````
A typo in "Japanese"

notxvilka commented on librz/include/rz_util/rz_str.h:
```diff
@@ -281,10 +290,44 @@ RZ_API RZ_OWN char *rz_str_stringify_raw_buffer(RzStrStringifyOpt *option, RZ_NU
+ * assert(rz_string_enc_same_char_width_as_utf8(RZ_STR_ENC_IBM290) == false);
+ *
+ * // ASCII character width is always one byte, and the equivalent
+ * // UTF-8 characters are also always 1 bytes.
```
"1 byte"

notxvilka commented on librz/include/rz_util/rz_str.h:
```diff
@@ -281,10 +290,44 @@ RZ_API RZ_OWN char *rz_str_stringify_raw_buffer(RzStrStringifyOpt *option, RZ_NU
 }
 
+RZ_API bool rz_string_enc_is_utf_native_endian(RzStrEnc enc);
+RZ_API size_t rz_string_enc_code_point_width(RzStrEnc enc);
```
Maybe an empty line between proto and inline function?

notxvilka commented on librz/include/rz_util/rz_utf16.h:
```diff
@@ -4,10 +4,13 @@
-RZ_API size_t rz_utf16le_decode(RZ_NONNULL const ut8 *buf, size_t buf_len, RZ_NONNULL RZ_OUT RzCodePoint *ch);
-RZ_API size_t rz_utf16be_decode(RZ_NONNULL const ut8 *buf, size_t buf_len, RZ_NONNULL RZ_OUT RzCodePoint *ch);
-RZ_API size_t rz_utf16le_encode(RZ_NONNULL RZ_OUT ut8 *buf, RzCodePoint ch);
+#define RZ_UTF16_FIRST_4BYTES_CODE_POINT 0x10000
```
Add a comment here like you did for UTF-32?

notxvilka commented on librz/include/rz_util/rz_utf8.h:
```diff
@@ -4,11 +4,13 @@
 /* For RzStrEnc definition */
 #include "rz_unicode.h"
 
+#define RZ_UTF8_CODE_POINT_WIDTH 1
```
Same? For consistency

notxvilka commented on librz/search/string_search.c:
```diff
@@ -215,11 +297,56 @@ RZ_API bool rz_search_collection_string_add(RZ_NONNULL RzSearchCollection *col,
+	}
+	if (!rz_string_code_points_align(ss->encoding, search_options->find_opts->alignment)) {
+		if (log_suggestions) {
+			RZ_LOG_INFO("The string encoding has code points of more than 1 byte. But setting search.align = 1.\n"
```
"But setting search.align = 1" should be reworded, e.g. "but search.align is set to 1" or something like that

notxvilka commented on librz/util/regex.c:
```diff
@@ -2,7 +2,7 @@
 
 #define PCRE2_STATIC
-#define PCRE2_CODE_UNIT_WIDTH 8
+#define PCRE2_CODE_UNIT_WIDTH 0
```
Is this on purpose?

Rot127 replied:
Yes, enables all supported libraries.

notxvilka commented on librz/util/regex.c:
```diff
@@ -166,9 +459,61 @@ RZ_API RzRegexStatus rz_regex_match(RZ_NONNULL const RzRegex *regex, RZ_NONNULL
+ * \param regex The regex pattern to match.
+ * \param text The text to search in. UTF-16 encoded in the host's endianness.
+ * \param text_size The length of the buffer pointed to by \p text.
+ * Can be set to RZ_REGEX_ZERO_TERMINATED if the buffer is a zero terminated string.
```
Could be indented to show it's a continuation of the previous line? Here and in other similar comments

notxvilka commented on librz/util/regex.c:
```diff
@@ -339,6 +675,228 @@ RZ_API RZ_OWN RzPVector /*<RzRegexMatch *>*/ *rz_regex_match_first(
+	rc = pcre2_match_16(regex, (PCRE2_SPTR16)text, text_size_code_units, text_offset_code_units, mflags | PCRE2_NO_UTF_CHECK, mdata, NULL);
+
+	if (rc == PCRE2_ERROR_NOMATCH) {
+		// Nothing matched return empty vector.
```
"Nothing matched, return an empty vector" or something like that.

notxvilka commented on librz/util/str.c:
```diff
@@ -2824,6 +2825,53 @@ RZ_API size_t rz_str_len_utf8(const char *s) {
+ *
+ * \return The number of Unicode code points *including* the final NUL.
+ */
+RZ_API size_t rz_str_utf8_num_ucp(RZ_NONNULL const char *str) {
```
If nonnull, please add an assert

notxvilka commented on librz/util/str.c:
```diff
@@ -2824,6 +2825,53 @@ RZ_API size_t rz_str_len_utf8(const char *s) {
+ *
+ * \return The number of bytes required for an UTF16 string, *including* the final NUL.
+ */
+RZ_API size_t rz_str_utf8_get_width_utf16(RZ_NONNULL const char *str) {
```
Same

notxvilka commented on librz/util/str.c:
```diff
@@ -3092,6 +3140,62 @@ RZ_API char *rz_str_utf16_decode(const ut8 *s, int len) {
+}
+
 // TODO: kill this completely, it makes no sense:
 RZ_API char *rz_str_utf16_encode(const char *s, int len) {
```
Maybe add here RZ_DEPRECATED or remove the comment if not relevant anymore?

notxvilka commented on librz/util/str.c:
```diff
@@ -4331,3 +4435,59 @@ RZ_API const char *rz_str_indent(int indent) {
 }
+
+/**
+ * \brief Checks given encoding if it is UTF-8, UTF16, or UTF-32
```
Make UTF-XX names uniform

notxvilka commented on librz/util/str_search.c:
```diff
@@ -230,8 +231,11 @@ static inline size_t buf_look_ahead(const RzUtilStrScanOptions *opt, RzStrEnc en
 	}
 }
 
+#define SCANNING_STACK_BUF_CHARS 16
```
Add a comment here what "scanning stack" means

notxvilka commented on librz/util/str_search.c:
```diff
@@ -304,9 +311,18 @@ static RzDetectedString *process_one_string(const ut8 *buf, const ut64 from, ut6
+			}
+			size_t mem_offset = needle;
+			byte_mem_map[utf8_char_offset] = mem_offset;
+			// printf("mem_char_map[0x%llx] = 0x%lx\n", utf8_char_offset, mem_offset);
```
Remove?

notxvilka commented on librz/util/str_search.c:
```diff
@@ -334,7 +350,8 @@ static RzDetectedString *process_one_string(const ut8 *buf, const ut64 from, ut6
 			char_count++;
 		} else {
-			/* \0 marks the end of C-strings */
+			/* \0 or undefined code point, marks the end of C-strings */
```
No need for comma

notxvilka commented on librz/util/str_search.c:
```diff
@@ -358,7 +375,14 @@ static RzDetectedString *process_one_string(const ut8 *buf, const ut64 from, ut6
 		ds->type = str_type;
 		ds->length = char_count;
 		ds->size = needle - str_addr;
+		if (stopped_with_undef_cp) {
```
Oh, a nasty bug spotted

Rot127 replied:
Made quite some strings a few bytes longer than they were actually.

notxvilka commented on librz/util/utf16.c:
```diff
@@ -66,39 +73,70 @@ RZ_API size_t rz_utf16_decode(RZ_NONNULL const ut8 *buf, size_t buf_len, RZ_NONN
- * \param codepoint The decoded code point.
+ * \param codepoint (Optional) The decoded code point.
+ * \param check_is_def If true, checks the code point against the defined
+ * Unicode table. It will not write \p cp and return 0 if the decoded code
```
Indentation?

notxvilka commented on librz/util/utf32.c:
```diff
@@ -9,13 +9,18 @@
- * Unicode code point was decoded.
+ * \param cp The decoded code point.
+ * \param check_is_def If true, checks the code point against the defined
+ * Unicode table. It will not write \p cp and return 0 if the decoded code
```
Indentation?

notxvilka commented on librz/util/utf32.c:
```diff
@@ -77,3 +82,31 @@ RZ_API bool rz_utf32_valid_code_point(RZ_NONNULL const ut8 *buf, size_t buf_len,
+		return 0;
+	}
+	if (big_endian) {
+		buf[3] = ucp & 0xff;
```
I think we had macroses for those

Rot127 replied:
Couldn't find them. But have to leave now. So can't search :(

notxvilka commented on librz/util/utf8.c:
```diff
@@ -369,12 +369,17 @@ RZ_API const char *rz_utf_block_name(int idx) {
- * Unicode code point was decoded.
+ * \param cp The decoded code point.
+ * \param check_is_def If true, checks the code point against the defined
+ * Unicode table. It will not write \p cp and return 0 if the decoded code
```
Indentation?

notxvilka commented on subprojects/packagefiles/pcre2/meson.build:
```diff
@@ -83,13 +81,25 @@ pcre2_includes = [
 
 pcre2_dep = declare_dependency(
-  link_with: libpcre2,
+  link_with: [ libpcre2_8, libpcre2_16, libpcre2_32 ],
```
Have you compared the size difference before and after? Just curious, we don't really have a choice...

Rot127 replied:
Three times the same library:

```
-rw-r--r--. 1 user user 8.0K Sep  6 11:04 libpcre2.a
```

```
-rw-r--r--. 1 user user 8.0K Sep  9 14:35 libpcre2_16.a
-rw-r--r--. 1 user user 8.1K Sep  9 14:35 libpcre2_32.a
-rw-r--r--. 1 user user 8.1K Sep  9 14:35 libpcre2_8.a
```

notxvilka reviewed (changes requested):
Overall is good, great changes! Mostly stylistic suggestions.

Rot127 committed:
Apply review comments.

notxvilka commented on librz/include/rz_util/rz_utf16.h:
```diff
@@ -4,10 +4,19 @@
+/**
+ * \brief Width of an UTF16 character in bytes.
+ */
+#define RZ_UTF16_CODE_POINT_WIDTH        2
```
clang-format error:
```
librz/include/rz_util/rz_utf16.h:14:34: error: code should be clang-formatted [-Wclang-format-violations]
#define RZ_UTF16_CODE_POINT_WIDTH        2
                                 ^
```

notxvilka reviewed (approved):
LGTM with one leftover
